### ISSUE TITLE
Cardboard SDK only supports armeabi-v7a

### ISSUE THREAD

**Issue by kanawish** [googlevr/gvr-android-sdk#38]
Since Android v0.5.5 core functionalities of the SDK are now implemented in C++.

The cardboard.jar only contains armeabi-v7a binaries.

Are there any plans to support v8 (nexus 5x, 6p) and other platforms? (Such as x86?) 

Also, would suggest if you opened up this (relatively simple) SDK to the community, you'd see a lot of positive contributions.

**Comment by dav-cz** [comment]
Hi Kanawish, we plan to ship binaries for other architectures in the next version.

As you mentioned, currently only armeabi-v7a binaries are shipped. Even though they might not be optimal, these should run well on arm v8 for sure and on x86 phones using emulation (at least it seemed pretty functional on the x86 phones we tested)

**Comment by kanawish** [comment]
Yesterday, I actually met someone at the AndroidTO conference who has ran into the same issue and found a fix. Apparently the problem can be solved by adding abiFilters to your build. I did *not* try this myself yet, that person shared the following gradle snippet:

```groovy
buildTypes {
   debug {
      ndk {
         abiFilters "armabi-v7a", "armabi" // Should do the trick.
      }
   }
} 
```

**Comment by kanawish** [comment]
Re-reading my original comment, I realize I didn't even explain that building for Nexus5x (armv8) failed, in the same manner as people have reported for x86. 

Happy to learn you'll be adding further native support :) thank you.

**Comment by dav-cz** [comment]
Just to confirm, when you say "building failed for Nexus 5X", do you mean the sample app from the SDK or your own code?  For me the sample app compiles and runs fine on Nexus 5X, with SDK 0.5.8.

If you mean your own code, and if you have your own native libraries, then you might have to apply abiFilters like you mentioned or copy all libraries into the same abi directory. (Though I've never tried that)

Let me know what your situation is, this way we could document that properly in a central place.

Thank you!

**Comment by kanawish** [comment]
Ah! I've made a mistake here. I went back and double checked with the 5X, and it is indeed my project that was the source of the issue, and I'm running into a totally different issue. I've managed to reproduce it in the sample code. It looks like a jni library inclusion bug. 

If I add an Android library module to the sample code, reconfigure the project to use the latest gradle plugin, (2.0.0-alpha1) and move the .jar into this module, I *actually* get the following at build time:

`Error:Execution failed for task ':mylibrary:transformNative_libsWithSyncJniLibsForRelease'.
> java.io.FileNotFoundException: /$HOME$/android-samples/cardboard-java/mylibrary/build/intermediates/bundles/release/jni/lib/armeabi-v7a/libvrtoolkit.so (No such file or directory)`

This should (?) normally work. Apparently this is a known issue:
https://github.com/nickbutcher/plaid/issues/45
https://code.google.com/p/android/issues/detail?id=193063

If I roll back to Gradle plugin 1.3.1, the build succeeds and indeed works on the 5X.

**Comment by kanawish** [comment]
Now, I've gone and rolled back my own personal project to use the 1.3.1 gradle plugin, and I am getting an unsatisfied link error (that's why I was originally so certain I was missing a arm64-v8a .so)

`java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.kanawish.androidvrtalk-1/base.apk"],nativeLibraryDirectories=[/data/app/com.kanawish.androidvrtalk-1/lib/arm64, /data/app/com.kanawish.androidvrtalk-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]] couldn't find "libvrtoolkit.so"
`

Full stack trace:

```
Process: com.kanawish.androidvrtalk, PID: 25399
 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.kanawish.androidvrtalk/com.kanawish.androidvrtalk.ui.VrTalkActivity}: android.view.InflateException: Binary XML file line #9: Binary XML file line #9: Error inflating class com.google.vrtoolkit.cardboard.CardboardView
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
     at android.app.ActivityThread.-wrap11(ActivityThread.java)
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
     at android.os.Handler.dispatchMessage(Handler.java:102)
     at android.os.Looper.loop(Looper.java:148)
     at android.app.ActivityThread.main(ActivityThread.java:5417)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
  Caused by: android.view.InflateException: Binary XML file line #9: Binary XML file line #9: Error inflating class com.google.vrtoolkit.cardboard.CardboardView
     at android.view.LayoutInflater.inflate(LayoutInflater.java:539)
     at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
     at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
     at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:393)
     at android.app.Activity.setContentView(Activity.java:2166)
     at com.kanawish.androidvrtalk.ui.VrTalkActivity.onCreate(VrTalkActivity.java:291)
     at android.app.Activity.performCreate(Activity.java:6237)
     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
     at android.app.ActivityThread.-wrap11(ActivityThread.java) 
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
     at android.os.Handler.dispatchMessage(Handler.java:102) 
     at android.os.Looper.loop(Looper.java:148) 
     at android.app.ActivityThread.main(ActivityThread.java:5417) 
     at java.lang.reflect.Method.invoke(Native Method) 
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
  Caused by: android.view.InflateException: Binary XML file line #9: Error inflating class com.google.vrtoolkit.cardboard.CardboardView
     at android.view.LayoutInflater.createView(LayoutInflater.java:645)
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764)
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704)
     at android.view.LayoutInflater.rInflate(LayoutInflater.java:835)
     at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
     at android.view.LayoutInflater.inflate(LayoutInflater.java:515)
     at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
     at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:393) 
     at android.app.Activity.setContentView(Activity.java:2166) 
     at com.kanawish.androidvrtalk.ui.VrTalkActivity.onCreate(VrTalkActivity.java:291) 
     at android.app.Activity.performCreate(Activity.java:6237) 
     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
     at android.app.ActivityThread.-wrap11(ActivityThread.java) 
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
     at android.os.Handler.dispatchMessage(Handler.java:102) 
     at android.os.Looper.loop(Looper.java:148) 
     at android.app.ActivityThread.main(ActivityThread.java:5417) 
     at java.lang.reflect.Method.invoke(Native Method) 
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
  Caused by: java.lang.reflect.InvocationTargetException
     at java.lang.reflect.Constructor.newInstance(Native Method)
     at android.view.LayoutInflater.createView(LayoutInflater.java:619)
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764) 
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704) 
     at android.view.LayoutInflater.rInflate(LayoutInflater.java:835) 
     at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:515) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
     at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:393) 
     at android.app.Activity.setContentView(Activity.java:2166) 
     at com.kanawish.androidvrtalk.ui.VrTalkActivity.onCreate(VrTalkActivity.java:291) 
     at android.app.Activity.performCreate(Activity.java:6237) 
     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
     at android.app.ActivityThread.-wrap11(ActivityThread.java) 
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
     at android.os.Handler.dispatchMessage(Handler.java:102) 
     at android.os.Looper.loop(Looper.java:148) 
     at android.app.ActivityThread.main(ActivityThread.java:5417) 
     at java.lang.reflect.Method.invoke(Native Method) 
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
  Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.kanawish.androidvrtalk-1/base.apk"],nativeLibraryDirectories=[/data/app/com.kanawish.androidvrtalk-1/lib/arm64, /data/app/com.kanawish.androidvrtalk-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]] couldn't find "libvrtoolkit.so"
     at java.lang.Runtime.loadLibrary(Runtime.java:367)
     at java.lang.System.loadLibrary(System.java:1076)
     at com.google.vrtoolkit.cardboard.CardboardViewNativeImpl.<init>(CardboardViewNativeImpl.java:103)
     at com.google.vrtoolkit.cardboard.ImplementationSelector.createCardboardViewApi(ImplementationSelector.java:35)
     at com.google.vrtoolkit.cardboard.CardboardView.init(CardboardView.java:905)
     at com.google.vrtoolkit.cardboard.CardboardView.<init>(CardboardView.java:241)
     at java.lang.reflect.Constructor.newInstance(Native Method) 
     at android.view.LayoutInflater.createView(LayoutInflater.java:619) 
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764) 
     at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704) 
     at android.view.LayoutInflater.rInflate(LayoutInflater.java:835) 
     at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:515) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
     at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
     at com.android.internal.policy.PhoneWindow.setContentView(PhoneWindow.java:393) 
     at android.app.Activity.setContentView(Activity.java:2166) 
     at com.kanawish.androidvrtalk.ui.VrTalkActivity.onCreate(VrTalkActivity.java:291) 
     at android.app.Activity.performCreate(Activity.java:6237) 
     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
     at android.app.ActivityThread.-wrap11(ActivityThread.java) 
     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
     at android.os.Handler.dispatchMessage(Handler.java:102) 
     at android.os.Looper.loop(Looper.java:148) 
     at android.app.ActivityThread.main(ActivityThread.java:5417) 
     at java.lang.reflect.Method.invoke(Native Method) 
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
```

**Comment by dav-cz** [comment]
Ok, so to summarize, if you use gradle 1.3.1, then
* you can build and run the sample SDK app on your N5X.
* you get an UnsatisfiedLinkError at run time for your own project, again on N5X.

Does your own project contain other .so libraries?

**Comment by powderluv** [comment]
Im using buck to build for arm64-v8a and my app fails to load with the same stack trace posted above. I think if the next cardboard.jar included the various other architectures it should work. Im going to try to stub the native lib to see if I can get it to work on arm64-v8a.

**Comment by kanawish** [comment]
@dav-cz I don't have any other .so in my project that I can see. (I don't have any of my own, and I took a quick look at my project dependencies, I did not see any other native libs.)

I'll try playing around with my project structure when I find some time, see if I can isolate the issue, or perhaps replicate it with the Cardboard SDK sample code.

**Comment by marktani** [comment]
I had the same issue on my OnePlus One.
* I could perfectly build and run the sample SDK on the phone. 
* I got a similar linker error during run time for my own project on the phone.

I fixed it by changing the gradle version for my own project from 1.3.0 to 1.5.0 and that's it.

**Comment by Chainfire** [comment]
Still no support for anything else than arm v7a with 0.6.0. If there's any real reason (compiler incompatibilities, hand-optimized assembler not easily ported, cthulhu being in a particularly bad mood, whatever), I'm sure we could all understand if you'd just share it.

**Comment by svenhenrik** [comment]
I get `INSTALL_FAILED_NO_MATCHING_ABIS` when trying to install the sample app on X86/X86_64 emulator. Does arm emulation work on the Emulator? If not, that native X86 support would be very nice. I'm not going to get all preachy, but making the whole SDK open source would also come be more than welcome...

Anyway, details: I'm using Android Studio 1.5.1 and a fresh clone of this repo. When I open the project I'm prompted to switch to Gradle 1.2.0, but it actually sets it to 1.5.0. Using 1.2.0 gives the same result. Apart from that I've made no changes. Tried running on emulator using X86 and X86_64, and install fails on both.

Adding the abiFilters snippet above to CardboardSample/build.gradle gives me the following error instead, and aborts installation:

`The currently selected variant "debug" uses split APKs, but none of the 1 split apks are compatible with the current device with density "-1" and ABIs "x86".`

**Comment by dav-cz** [comment]
One problem we have is that our native library is folded into a jar, which doesn't play well with gradle's abiFilters. Therefore having libs for multiple architectures into the jar could break applications that compile native code for only a subset of architectures, when they run on phones with cpus the application didn't compile for.
For the next release we plan to ship the SDK as an AAR, which will overall greatly help with this issue. Thanks for your patience!

**Comment by jclova** [comment]
Hi @kanawish,
I am the guy whom you met in AndroidTO.
Also the same person who posted the solution to the stackoverflow.

```
buildTypes {
   release { 
      ndk {
         abiFilters "armabi-v7a", "armabi" // Should do the trick.
      }
   }
   debug { 
      ndk {
         abiFilters "armabi-v7a", "armabi" // Should do the trick.
      }
   }
} 
```

Above code made it work for me for all Intel devices (Zenfone 5) & 64-bit devices (S6, Note 5).

However, like @kanawish mentioned, when I build a library that uses CardboardSDK, then it doesn't work. I get a compile time error: 
```
Error:Execution failed for task ':(project):transformNative_libsWithSyncJniLibsForRelease'.
> java.io.FileNotFoundException: /(path)/build/intermediates/bundles/release/jni/lib/armeabi-v7a/libvrtoolkit.so (No such file or directory)

```

You can easily reproduce this:
1.  Create an empty project.
2.  Create an Android library module.
3. Add cardboard.jar to the libarary/lib folder.
4. Add "compile fileTree(dir: 'libs', include: ['*.jar'])" to the library build.gradle.
5. Press: "Sync Project with Gradle Files" button.

For convenience, I attached the project for you to easily reproduce this issue.
https://drive.google.com/open?id=0BygTefPD845LbERwcTFqakR1aDg

I hope aar solves this issue.


[CardboardSDKTest2.zip](https://github.com/googlesamples/cardboard-java/files/110537/CardboardSDKTest2.zip)

**Comment by kanawish** [comment]
@jclova and others that might be interested, (while waiting on the new .aar structure Cardboard SDK) I seem to have a solution specifically for the 'lib build' problem in my own library. The hack is to extract the `libvrtoolkit.so` file from the cardboard .jar, and move it into my library module under `myModule/main/jniLibs/armeabi-v7a/libvrtoolkit.so`.

I still include the cardboard.jar under `myModule/lib` as I would otherwise.

**Comment by jazzyjester** [comment]
Hello 
I tried @kanawish solution in eclipse and it didn't worked.
I have error like that in 0.6.0 cardboard.jar version

[2016-02-22 17:06:52 --Default] The library 'cardboard.jar' contains native libraries that will not run on the device.
[2016-02-22 17:06:52 --Default] Additionally some of those libraries will interfer with the installation of the application because of their location in lib/
[2016-02-22 17:06:52 --Default] lib/ is reserved for NDK libraries.
[2016-02-22 17:06:52 --Default] The following libraries were found:
[2016-02-22 17:06:52 --Default]  - lib/armeabi-v7a/libvrtoolkit.so

**Comment by kanawish** [comment]
Hello @jazzyjester. I've personally been off Eclipse since pre-Android-Studio days, so I can't really speak to that setup or the build chain on an Eclipse setup.

As of writing, the only way I got my example code running _cross-platform_ under the new build system was by moving away from using library modules (as per above discussion). The hack above worked for `arm-v7a` devices, but the final `.APK` crashes at native lib load time.

If you're curious, check out [my un-modularized code sample](https://github.com/kanawish/AndroidVRTalk/tree/feature/unmodularize). It's messy, but for my purposes (an Intro-to-Cardboard conference talk) it does the job.

**Comment by jclova** [comment]
Thank you @kanawish, your solution works like a charm.

**Comment by jclova** [comment]
Also confirming @kanawish's ADDENDUM above.
Non armv7 devices don't work. Reverting back to the old version that doesn't use libvrtoolkit.so.
I will wait for AAR.

**Comment by ph0b** [comment]
I've seen the AAR has been released with v0.7.0 but it only includes armeabi-v7a architecture.
arm64-v8a and x86 libraries are available separately here: https://github.com/googlesamples/cardboard-java/tree/master/libraries/native

Why not including these by default in the AAR?
It's much less error-prone for developers to filter them out when needed, using abiFilter for example, than having to reinclude them manually outside of the gradle dependency system, after having noticed crashes such as the ones related above.

Having armeabi-v7a only aars causes a lot of troubles when the rest of your codebase supports other architectures. Can you please improve this and support a maximum of architectures directly within the .aar ? btw, this issue is quite related to this improvement I've proposed regarding lint-checks about the handling of dependency with native components: https://code.google.com/p/android/issues/detail?id=94805

**Comment by sigmaxipi** [comment]
@ph0b : including the other architectures by default means people might accidentally include other architectures if their build process isn't configured correctly. This would bloat their app and they might not notice. Requiring the extra step of explicitly including alternative architectures avoids this problem. 

There is no perfect solution that supports all the use cases with zero extra configuration.

**Comment by ph0b** [comment]
both ways will lead apps that aren't configured correctly, with a consistent set of libs for each architecture, to crash. It's not just bloat.

In any case, it's a lot better to let people use abiFilter to filter architectures if they want, which is truly 1 step that is going to survive across version upgrades, than the other way around that requires 4 manual steps for each new releases - which is going to lead to even more crashes when people will upgrade the .aar and not the additional libs.

**Comment by isometriq** [comment]
Here's a follow-up ..I was able to run the latest GVR lib on the emulator. I've answered my issue here, with some instructions:
https://github.com/googlevr/gvr-android-sdk/issues/138

I've only used the trick that was proposed and discussed here ..so thanks for that.

**Comment by sigmaxipi** [comment]
We've switched to using multiple architectures in the .aars
